rfcs/20190729-keras-preprocessing.md

```diff
@@ -0,0 +1,738 @@
+        class_names: Only valid if "targets" is "inferred". This is the explict
+            list of class names (must match names of subdirectories). Used
+            to control the order of the classes (otherwise alphanumerical order is used).
+        color_mode: One of "grayscale", "rgb", "rgba". Default: "rgb".
+            Whether the images will be converted to
+            have 1, 3, or 4 channels.
+        batch_size: Size of the batches of data (default: 32).
+        shuffle: Whether to shuffle the data (default: True)
+            If set to False, sorts the data in alphanumeric order.
+        seed: Optional random seed for shuffling and transformations.
+        follow_links: Whether to follow links inside
+            subdirectories (default: False).
+        validation_split: Optional float between 0 and 1,
+            fraction of data to reserve for validation.
+        subset: One of "training" or "validation". Only used if `validation_split` is set.
+    """
+```
+
+Notes:
+- Resizing (`target_size`) is moved to constructor (`resize`).
+- `validation_split` is moved from constructor to here.
+
+#### Question: how to support image segmentation in a simple way?
+
+**Requirements:**
+- Image loading and image augmentation should be synced across inputs and targets
+- It should be possible to use different standardization preprocessing (outside of augmentation) across inputs and targets
+
+**Proposal:** (TBD)
+
+```python
+augmenter = ImageDataGenerator(...)
+input_preprocessor = ImageDataGenerator(...)  # or write your own function
+target_preprocessor = ImageDataGenerator(...)  # or write your own function
+
+input_ds = augmenter.dataset_from_directory(input_dir, targets=None, seed=123)
+target_ds = augmenter.dataset_from_directory(target_dir, targets=None, seed=123)  # target_ds should have same structure.
+input_ds = input_ds.map(input_preprocessor.standardize)  # `standardize` could work on tensors.
+target_ds = input_ds.map(target_preprocessor.standardize)
+
+ds = tf.data.Dataset.zip((input_ds, target_ds))
+
+model.fit(ds)
+```
+
+
+
+#### Dataframe methods
+
+```python
+# CURRENT
+
+def flow_from_dataframe(
+    self,
+    dataframe,
+    directory=None,
+    x_col='filename',
+    y_col='class',
+    weight_col=None,
+    target_size=(256, 256),
+    color_mode='rgb',
+    classes=None,
+    class_mode='categorical',
+    batch_size=32,
+    shuffle=True,
+    seed=None,
+    save_to_dir=None,
+    save_prefix='',
+    save_format='png',
+    subset=None,
+    interpolation='nearest',
+    validate_filenames=True,
+    **kwargs):
+```
+
+```python
+# PROPOSED
+
+def generator_from_dataframe(
+    self,
+    dataframe,
+    directory=None,
+    data_column='filename',
+    target_column='class',
+    weight_column=None,
+    color_mode='rgb',
+    class_mode='categorical',
+    batch_size=32,
+    shuffle=True,
+    seed=None,
+    subset=None):
+
+def dataset_from_dataframe(
```

Comment:
Is there a world where we don't need all of these, and we can just determine data type from the passed object? String is dir path, array/dataframe can be type-checked. Then we just have to_dataset(), without the need for so many separate APIs.

Reply:
That would be a simplification in terms of the set of methods available, but the signatures would get quite complex / confusing since each action has its own specific keyword arguments on top of the shared ones (e.g. `follow_links`, `target_mode`...)